sycl-fusion/passes/target/TargetFusionInfo.cpp

```diff
@@ -0,0 +1,371 @@
+    SmallPtrSet<Constant *, 8> DeletedFuncs{Funcs.begin(), Funcs.end()};
+    SmallVector<MDNode *> ValidKernels;
+    auto *OldAnnotations = LLVMMod->getNamedMetadata("nvvm.annotations");
+    for (auto *Op : OldAnnotations->operands()) {
+      if (auto *TOp = dyn_cast<MDTuple>(Op)) {
+        if (auto *COp = dyn_cast_if_present<ConstantAsMetadata>(
+                TOp->getOperand(0).get())) {
+          if (!DeletedFuncs.contains(COp->getValue())) {
+            ValidKernels.push_back(Op);
+            // Add to the set to also remove duplicate entries.
+            DeletedFuncs.insert(COp->getValue());
+          }
+        }
+      }
+    }
+    LLVMMod->eraseNamedMetadata(OldAnnotations);
+    auto *NewAnnotations =
+        LLVMMod->getOrInsertNamedMetadata("nvvm.annotations");
+    for (auto *Kernel : ValidKernels) {
+      NewAnnotations->addOperand(Kernel);
+    }
+  }
+
+  void addKernelFunction(Function *KernelFunc) const override {
+    auto *NVVMAnnotations =
+        LLVMMod->getOrInsertNamedMetadata("nvvm.annotations");
+    auto *MDOne = ConstantAsMetadata::get(
+        ConstantInt::get(Type::getInt32Ty(LLVMMod->getContext()), 1));
+    auto *MDKernelString = MDString::get(LLVMMod->getContext(), "kernel");
+    auto *MDFunc = ConstantAsMetadata::get(KernelFunc);
+    SmallVector<Metadata *, 3> KernelMD({MDFunc, MDKernelString, MDOne});
+    auto *Tuple = MDTuple::get(LLVMMod->getContext(), KernelMD);
+    NVVMAnnotations->addOperand(Tuple);
+  }
+
+  ArrayRef<StringRef> getKernelMetadataKeys() const override {
+    // FIXME: Check whether we need to take care of sycl_fixed_targets.
+    static SmallVector<StringRef> Keys{{"kernel_arg_buffer_location",
+                                        "kernel_arg_runtime_aligned",
+                                        "kernel_arg_exclusive_ptr"}};
+    return Keys;
+  }
+
+  void createBarrierCall(IRBuilderBase &Builder,
+                         int BarrierFlags) const override {
+    if (BarrierFlags == -1) {
+      return;
+    }
+    // Emit a call to llvm.nvvm.barrier0. From the user manual of the NVPTX
+    // backend: "The ‘@llvm.nvvm.barrier0()’ intrinsic emits a PTX bar.sync 0
+    // instruction, equivalent to the __syncthreads() call in CUDA."
+    Builder.CreateIntrinsic(Intrinsic::NVVMIntrinsics::nvvm_barrier0, {}, {});
+  }
+
+  // Corresponds to the definitions in the LLVM NVPTX backend user guide:
+  // https://llvm.org/docs/NVPTXUsage.html#address-spaces
+  unsigned getPrivateAddressSpace() const override { return 0; }
+  unsigned getLocalAddressSpace() const override { return 3; }
+
+private:
+  using TargetFusionInfoImpl::TargetFusionInfoImpl;
+};
+
+//
+// TargetFusionInfo
+//
+
+TargetFusionInfo::TargetFusionInfo(llvm::Module *Mod) {
+  llvm::Triple Tri(Mod->getTargetTriple());
+  if (Tri.isNVPTX()) {
+    Impl =
+        std::shared_ptr<NVPTXTargetFusionInfo>(new NVPTXTargetFusionInfo(Mod));
```

Comment:
```suggestion
    Impl =
        std::shared_ptr<NVPTXTargetFusionInfo>(new NVPTXTargetFusionInfo(Mod));
```
Use `std::make_shared` here and below

Reply:
That was initially not possible due to `protected`/`private` constructors. After moving the implementation classes to the `.cpp` file, we can now have public constructors and use `make_shared`.